[BB-2088] Add %%COURSE_ID%% to html block

Agrendalath opened the pull request:
This adds support for replacing `%%COURSE_ID%%` keyword in built-in HTML block.

**JIRA tickets**: [OSPR-4115](https://openedx.atlassian.net/browse/OSPR-4115)

**Sandbox URL**: https://pr23174.sandbox.opencraft.hosting/

**Merge deadline**: None

**Testing instructions**:
1. Create new HTML block.
1. Place the following content inside:
   ```
   User ID: %%USER_ID%%
   Course ID: %%COURSE_ID%%
   ```
1. Confirm that both IDs have been replaced.

**Sandbox testing instructions**:
1. Log in via the [admin page](https://pr23174.sandbox.opencraft.hosting/admin/). The DB is missing `email_marketing_emailmarketingconfiguration` table, but running the migrations did not fix the problem, so I didn't debug it, as it's not related.
1. Go to [LMS](https://pr23174.sandbox.opencraft.hosting/courses/course-v1:edX+DemoX+Demo_Course/courseware/9fca584977d04885bc911ea76a9ef29e/07bc32474380492cb34f76e5f9d9a135/1?activate_block_id=block-v1%3AedX%2BDemoX%2BDemo_Course%2Btype%40vertical%2Bblock%4045c7cedb4bfe46f4a68c78787151cfb5) and [Studio](https://studio.pr23174.sandbox.opencraft.hosting/container/block-v1:edX+DemoX+Demo_Course+type@vertical+block@45c7cedb4bfe46f4a68c78787151cfb5) and confirm that the block has the keywords replaced.

**Reviewers**
- [x] @xitij2000  
- [ ] edX reviewer[s] TBD

edx-webhook commented:
Thanks for the pull request, @Agrendalath! I've created [OSPR-4115](https://openedx.atlassian.net/browse/OSPR-4115) to keep track of it in JIRA. JIRA is a place for product owners to prioritize feature reviews by the engineering development teams.

Feel free to add as much of the following information to the ticket:
- supporting documentation
- edx-code email threads
- timeline information ("this must be merged by XX date", and why that is)
- partner information ("this is a course on edx.org")
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will still be done via the GitHub pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).

natabene commented:
@Agrendalath Thank you for your contribution. Please let me know once this is ready for our review.

ormsbee commented on common/djangoapps/util/keyword_substitution.py:
```diff
@@ -46,14 +47,19 @@ def substitute_keywords(string, user_id, context):
     KEYWORD_FUNCTION_MAP = {
         '%%USER_ID%%': lambda: anonymous_id_from_user_id(user_id),
         '%%USER_FULLNAME%%': lambda: context.get('name'),
+        '%%COURSE_ID%%': lambda: context.course_id.html_id(),
```
@bradenmacdonald: Will this break under the Blockstore runtime?

bradenmacdonald replied:
No, [it won't break thanks to `shims.py`](https://github.com/edx/edx-platform/blob/2eeae3b9ad07fdff2cc87ae67f0f651332950443/openedx/core/djangoapps/xblock/runtime/shims.py#L349-L359), but I consider calling `block.runtime.course_id` deprecated and would prefer to see `block.scope_ids.usage_id.context_key` used instead.

bradenmacdonald replied:
It might break in the email context though, if `context.course_id` is an AttributeError.

bradenmacdonald commented on common/lib/xmodule/xmodule/html_module.py:
```diff
@@ -117,9 +119,7 @@ def student_view_data(self, context=None):  # pylint: disable=unused-argument
-        if self.data is not None and getattr(self.system, 'anonymous_student_id', None) is not None:
-            return self.data.replace("%%USER_ID%%", self.system.anonymous_student_id)
-        return self.data
+        return substitute_keywords_with_data(self.data, self.system)
```
Won't this also add several other new %%field%% substitutions, like `%%COURSE_END_DATE%%` ?

I'd prefer a simple change like the following. This is (AFAIK) the most current way to write this code, not using old modulesystem properties that I consider deprecated:
```
if self.data is not None:
    data = self.data
    data = data.replace("%%USER_ID%%", getattr(self.runtime, 'anonymous_student_id', "%%USER_ID%%"))
    data = data.replace("%%COURSE_ID%%", block.scope_ids.usage_id.context_key)
    return data
```

This also avoids reaching into `common.djangoapps.util` (built-in XBlocks should try to reduce dependencies on other parts of edx-platform IMHO)

Agrendalath replied:
This was my first approach, but we ended proposing the alternative one. I changed this according to your review. Thank you!

Agrendalath commented on common/djangoapps/util/tests/test_keyword_sub_utils.py:
```diff
@@ -51,6 +52,34 @@ def test_course_name_sub(self, test_string, expected):
+        """ Tests subbing course ID."""
+        test_string = "Turn %%COURSE_ID%% into course HTML id"
+        mock_context = Mock()
+        mock_context.course_id = CourseKey.from_string('/'.join(['org', 'course', 'run']))
```
Sure, I changed this to `CourseLocator` in the other test.

Agrendalath commented:
@natabene, @ormsbee, this is ready for the review.

natabene commented:
@Agrendalath Great, thanks for letting me know. I will line it up for our review. Please work with @giovannicimolin to add this to prioritization list.

giovannicimolin commented:
@Agrendalath I've added this to the OSPR prioritization document.

bradenmacdonald commented on common/lib/xmodule/xmodule/html_module.py:
```diff
@@ -116,8 +116,12 @@ def student_view_data(self, context=None):  # pylint: disable=unused-argument
+            data = self.data
+            if getattr(self.runtime, 'anonymous_student_id', None):
+                data = data.replace("%%USER_ID%%", self.runtime.anonymous_student_id)
+            data = data.replace("%%COURSE_ID%%", six.text_type(self.scope_ids.usage_id.context_key))
```
Nit: we have python 3 now so you can use `str()` instead of `six.text_type()`

bradenmacdonald commented on common/lib/xmodule/xmodule/html_module.py:
```diff
@@ -459,10 +463,9 @@ def get_html(self):
-            if self.system.anonymous_student_id:
-                return self.data.replace("%%USER_ID%%", self.system.anonymous_student_id)
-            return self.data
+        data = HtmlBlock.get_html(self)
```
Does `super().get_html()` not work here?

Agrendalath replied:
It does, good catch!

bradenmacdonald commented on common/lib/xmodule/xmodule/tests/test_html_module.py:
```diff
@@ -2,10 +2,11 @@
 
 import unittest
 
+import six
```
Same: don't use `six` anymore.

bradenmacdonald reviewed (approved):
:+1: please just addess the py3 nits and then squash+rebase.

- [x] I tested this: on the sandbox - thanks!
- [x] I read through the code
- [x] I checked for accessibility issues: n/a
- [x] Includes documentation - no, it would be good to document this and `%%USER_ID%%` at https://edx.readthedocs.io/projects/open-edx-building-and-running-a-course/en/latest/course_components/create_html_component.html but that shouldn't block this PR

Agrendalath commented:
@bradenmacdonald, I addressed the nits.

> Includes documentation - no, it would be good to document this and %%USER_ID%% at https://edx.readthedocs.io/projects/open-edx-building-and-running-a-course/en/latest/course_components/create_html_component.html but that shouldn't block this PR

Good point, I'll open a PR at the beginning of the next sprint. I see that the `%%USER_ID%%` is already documented [here](https://edx.readthedocs.io/projects/edx-partner-course-staff/en/latest/exercises_tools/qualtrics.html#create-the-html-component-in-studio), though it's a pretty narrow use-case. Should we expand that section to include `%%COURSE_ID%%` too?

bradenmacdonald commented:
FYI: I just noticed #24930 from @swalladge. I don't think there's any conflict here, but that one will need to be rebased after this merges. We could also incorporate that fix now if you prefer, but then I'll need to re-test and it will delay getting this merged slightly.

bradenmacdonald commented:
> Good point, I'll open a PR at the beginning of the next sprint. I see that the %%USER_ID%% is already documented here, though it's a pretty narrow use-case. Should we expand that section to include %%COURSE_ID%% too?

I don't think it would be too useful in that narrow section, probably better to add it to https://edx.readthedocs.io/projects/open-edx-building-and-running-a-course/en/latest/course_components/create_html_component.html

Agrendalath commented:
> FYI: I just noticed #24930 from @swalladge. I don't think there's any conflict here, but that one will need to be rebased after this merges. We could also incorporate that fix now if you prefer, but then I'll need to re-test and it will delay getting this merged slightly.

@bradenmacdonald, I'm fine with both approaches, though merging this first would be preferable, as it keeps these PRs' contexts separated (and preserves the chronological order).
However I already resolved these conflicts during creating the client's branch, because the changes from #24930 were already present in our common branch. Therefore this is not a strong opinion, as we won't need to do anything around this in Juniper anymore.

edx-webhook commented:
@Agrendalath 🎉 Your pull request was merged!

Please take a moment to [answer a two question survey](https://docs.google.com/forms/d/e/1FAIpQLSceJOyGJ6JOzfy6lyR3T7EW_71OWUnNQXp68Fymsk3MkNoSDg/viewform?usp=pp_url&entry.1671973413=edx/edx-platform&entry.867055334=https://github.com/edx/edx-platform/pull/23174&entry.1484655318=https://github.com/Agrendalath&entry.752974735=2020-02-24+11:29&entry.1917517419=2020-10-07+16:36&entry.2133058324=Yes) so we can improve your experience in the future.
